src/NUnitEngine/nunit.engine.core/Drivers/NUnitFrameworkApi2018.cs

```diff
@@ -0,0 +1,293 @@
+            {
+                throw new NUnitEngineException($"Requested runtime {requestedRuntime} is not suitable for use with test assembly {_testAssemblyPath}", ex);
+            }
+            catch (SerializationException ex)
+            {
+                throw new NUnitEngineException("The NUnit 3 driver cannot support this test assembly. Use a platform specific runner.", ex);
+            }
+            catch (Exception ex) 
+            {
+                string msg = $"Failed to load {_nunitRef.FullName}\r\n  Codebase: {_nunitRef.CodeBase}";
+                throw new Exception(msg, ex);
+            }
+
+            _frameworkControllerType = _frameworkController?.GetType();
+            log.Debug($"Created FrameworkController {_frameworkControllerType?.Name}");
+
+            var controllerAssembly = _frameworkControllerType?.Assembly?.GetName();
+            log.Debug($"Controller assembly is {controllerAssembly}");
+#else
+            _assemblyLoadContext = new TestAssemblyLoadContext(testAssemblyPath);
+
+            _testAssembly = LoadAssembly(testAssemblyPath);
+            _frameworkAssembly = LoadAssembly(_nunitRef);
+
+            _frameworkController = CreateInstance(CONTROLLER_TYPE, _testAssembly, idPrefix, settings);
+            if (_frameworkController == null)
+            {
+                log.Error(INVALID_FRAMEWORK_MESSAGE);
+                throw new NUnitEngineException(INVALID_FRAMEWORK_MESSAGE);
+            }
+#endif
+
+            _frameworkControllerType = _frameworkController?.GetType();
+            log.Debug($"Created FrameworkController {_frameworkControllerType?.Name}");
+
+            log.Debug($"Loaded {testAssemblyPath}");
+            return (string)ExecuteMethod(LOAD_METHOD);
+        }
+
+        public int CountTestCases(string filter)
+        {
+            CheckLoadWasCalled();
+            object? count = ExecuteMethod(COUNT_METHOD, filter);
+            return count != null ? (int)count : 0;
+        }
+
+        public string Run(ITestEventListener? listener, string filter)
+        {
+            CheckLoadWasCalled();
+            log.Info("Running {0} - see separate log file", Path.GetFileName(_testAssemblyPath.ShouldNotBeNull()));
+            Action<string>? callback = /*listener != null ? listener.OnTestEvent :*/ (Action<string>?)null;
+            return (string)ExecuteMethod(RUN_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+        }
+
+        public void RunAsync(Action<string> callback, string filter)
+        {
+            CheckLoadWasCalled();
+            log.Info("Running {0} - see separate log file", Path.GetFileName(_testAssemblyPath.ShouldNotBeNull()));
+            ExecuteMethod(RUN_ASYNC_METHOD, new[] { typeof(Action<string>), typeof(string) }, callback, filter);
+        }
+
+        public void StopRun(bool force)
+        {
+            ExecuteMethod(STOP_RUN_METHOD, force);
```

Comment:
No call to `CheckLoadWasCalled`. Preventable by moving the check to the `Execute` method and deleting it from all those places that call it just before calling `Execute`

Reply:
It was present long ago, but StopRun does nothing if no test is running. So CheckLoadWasCalled would sometimes give a false error if the run had already terminated.